qualtran/qir_example.ipynb

```diff
@@ -0,0 +1,685 @@
+    "    n: int\n",
+    "\n",
+    "    @property\n",
+    "    def signature(self):\n",
+    "        return Signature([\n",
+    "            Register('x', bitsize=1, shape=(self.n,)),\n",
+    "            Register('y', bitsize=1, shape=(self.n,)),\n",
+    "        ])\n",
+    "\n",
+    "    def build_composite_bloq(\n",
+    "            self, bb: BloqBuilder, *, x: SoquetT, y: SoquetT\n",
+    "    ) -> Dict[str, SoquetT]:\n",
+    "        for i in range(self.n):\n",
+    "            x[i], y[i] = bb.add(ExampleBaseBloq(), x=x[i], y=y[i])\n",
+    "        return {'x': x, 'y': y}"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "outputs": [],
+   "source": [
+    "PYQIR_OP_MAP = {\n",
+    "    # Single-Qubit Clifford Gates\n",
+    "    Hadamard: pyqir._native.h,\n",
+    "    XGate: pyqir._native.x,\n",
+    "    YGate: pyqir._native.y,\n",
+    "    ZGate: pyqir._native.z,\n",
+    "    # Single-Qubit Rotation Gates\n",
+    "    Rx: pyqir._native.rx,\n",
+    "    Ry: pyqir._native.ry,\n",
+    "    Rz: pyqir._native.rz,\n",
+    "    # Single-Qubit Non-Clifford Gates\n",
+    "    TGate: pyqir._native.t,\n",
+    "    # Two-Qubit Gates\n",
+    "    TwoBitSwap: pyqir._native.swap,\n",
+    "    CNOT: pyqir._native.cx,\n",
+    "    # Three-Qubit Gates\n",
+    "    Toffoli: pyqir._native.ccx,\n",
+    "}"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 110,
+   "outputs": [],
+   "source": [
+    "def get_num_qubits_for_bloq(bloq: qualtran.Bloq):\n",
+    "    num_qubits = 0\n",
+    "    for register in bloq.signature.lefts():\n",
+    "        shape = register.shape[0] if len(register.shape) != 0 else 1\n",
+    "        num_qubits += register.bitsize*shape\n",
+    "    return num_qubits\n",
+    "\n",
+    "def create_func_for_bloq(bloq: qualtran.Bloq, name, qubit_type, void_type, mod: pyqir.Module):\n",
+    "    num_qubits = get_num_qubits_for_bloq(bloq)\n",
+    "    return Function(\n",
+    "    pyqir.FunctionType(void_type, [qubit_type]*num_qubits),\n",
+    "    Linkage.EXTERNAL,\n",
+    "    name,\n",
+    "    mod\n",
+    "    )\n",
+    "\n",
+    "def create_ir_map(bloq: qualtran.Bloq):\n",
+    "    param_counter = 0\n",
+    "    ir_map = {}\n",
+    "    # Loop through all registers in signature\n",
+    "    for register in bloq.signature.lefts():\n",
+    "        shape = register.shape[0] if len(register.shape) != 0 else 1 # get the shape as an int (we are asumming its 1d for simplicity)\n",
+    "        # map the (register_name, index_in_register) to the overall index\n",
+    "        ir_map.update({(register.name, i*register.bitsize + j): param_counter + i*register.bitsize + j for i in range(shape) for j in range(register.bitsize)})\n",
+    "        param_counter += register.bitsize * shape\n",
+    "    return ir_map\n",
```

Comment:
We can move all the utility functions to a new file `bloq_to_qir.py` and add tests using the example cases shown above.